airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/demo/file_based/iface/StreamDestination.java

```diff
@@ -0,0 +1,20 @@
+package io.airbyte.integrations.destination.snowflake.demo.file_based.iface;
+
+import java.io.File;
+
+public interface StreamDestination extends AutoCloseable {
+
+  /**
+   * Do any setup prior to writing data. For example, setting up destination tables, etc.
+   */
+  void setup() throws Exception;
+
+  /**
+   * Commit the data in this file to the destination. This method may consume additional
+   * memory proportional to the size of the incoming file and all uncommitted files,
+   * and/or the number of uncommitted files.
+   * <p>
+   * This method <b>MUST</b> be thread-safe.
+   */
+  void upload(String file, int numRecords, int numBytes) throws Exception;
```

Comment:
I landed on just using local files as the interface. Having multiple storage mediums would be _nice_, but it's definitely more complicated. Local file IO is generally fast enough that I don't feel too bad about this.

Managing records directly is a PITA that I think is best handled in "platform" code. Then DB/DW destinations can just plug in the right RecordWriter, and use the appropriate `PUT`/`COPY` commands. This also makes memory management a lot easier, since we don't need to mess around with different types of OutputStreams/S3Writer/whatever.

Reply:
> Managing records directly is a PITA that I think is best handled in "platform" code.

This is a good guiding principal for this work!  We know that there are funny representations of records (fully parsed, partially parsed), and not needing to think about that in each destination would be nice.